src/plugins/keras_support/categorical_crossentropy_operation.cpp

```diff
@@ -0,0 +1,291 @@
+        });
+        vector_type ans = sum2d_axis1(target_);
+        output.matrix() = std::move(output_);
+        return primitive_argument_type{ ans };
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+#if defined(PHYLANX_HAVE_BLAZE_TENSOR)
+     blaze::DynamicTensor<double> cat_cross_operation::softmax3d_axis2(
+        arg_type& arg) const
+    {
+        auto t = arg.tensor();
+
+        blaze::DynamicTensor<double> result(t.pages(), t.rows(), t.columns());
+        for (std::size_t i = 0; i != t.pages(); ++i)
+        {
+            auto slice = blaze::pageslice(t, i);
+            blaze::pageslice(result, i) = blaze::softmax<blaze::rowwise>(slice);
+        }
+        return result;
+    }
+
+    primitive_argument_type cat_cross_operation::cat_cross3d(
+        arg_type&& target, arg_type&& output, bool from_logits) const
+    {
+        tensor_type&& output_ = output.tensor();
+        tensor_type&& target_ = target.tensor();
+        if(from_logits)
+        {
+            output_ = softmax3d_axis2(output);
+        }
+        else
+        {
+            auto norm = sum3d_axis2(output_);
+            for(std::size_t j = 0; j < output_.columns(); ++j) {
+                auto slice = blaze::columnslice(output_,j);
+                slice = blaze::map(slice, norm,[](double s_,double n_){
+                    return s_/n_;
+                });
+            }
+        }
+
+        target_ = blaze::map(target_, output_,[](double t_, double o_){
+            return -t_*std::log((std::min)(clip_high,(std::max)(clip_low,o_)));
+        });
+        matrix_type ans = sum3d_axis2(target_);
+        output.tensor() = std::move(output_);
+        return primitive_argument_type{ ans };
+    }
+#endif
+
+    ///////////////////////////////////////////////////////////////////////////
+    hpx::future<primitive_argument_type> cat_cross_operation::eval(
+        primitive_arguments_type const& operands,
+        primitive_arguments_type const& args,
+        eval_context ctx) const
+    {
+        for (auto const& i : operands)
+        {
+            if (!valid(i))
+            {
+                HPX_THROW_EXCEPTION(hpx::bad_parameter,
+                    "cat_cross_operation::eval",
+                    util::generate_error_message(
+                        "the cat_cross_operation primitive requires that the "
+                        "arguments given by the operands array are "
+                        "valid",
+                        name_, codename_));
+            }
+        }
+
+        auto this_ = this->shared_from_this();
+        return hpx::dataflow(hpx::launch::sync,
+            hpx::util::unwrapping([this_ = std::move(this_)](
+                                      primitive_arguments_type&& args)
+                                      -> primitive_argument_type {
+                // Extract logits
+                std::int64_t from_logits =
```

Comment:
Why do you use `std::int64_t` as the type of `from_logits`, wouldn't a simple `bool` work equally well?

Reply:
It was a typo, but it compiled and ran so I didn't notice. :)